packages/opentelemetry-tracer/src/instrumentation/PluginLoader.ts

```diff
@@ -0,0 +1,148 @@
+   * function will attach a hook to be called the first time the module is
+   * loaded.
+   * @param modulesToPatch A list of plugins.
+   */
+  loadPlugins(modulesToPatch: string[]): PluginLoader {
+    if (this._hookState === HookState.UNINITIALIZED) {
+      const plugins = modulesToPatch.reduce(
+        (plugins: PluginNames, moduleName: string) => {
+          plugins[moduleName] = utils.defaultPackageName(moduleName);
+          return plugins;
+        },
+        {} as PluginNames
+      );
+      const modulesToHook = Object.keys(plugins);
+      // Do not hook require when no module is provided. In this case it is
+      // not necessary. With skipping this step we lower our footprint in
+      // customer applications and require-in-the-middle won't show up in CPU
+      // frames.
+      if (modulesToHook.length === 0) {
+        this._hookState = HookState.DISABLED;
+        return this;
+      }
+
+      // Enable the require hook.
+      hook(modulesToHook, (exports, name, baseDir) => {
+        if (this._hookState !== HookState.ENABLED) return exports;
+
+        const moduleName = plugins[name];
+        // Get the module version.
+        const version = utils.getPackageVersion(this.logger, baseDir as string);
+        this.logger.info(
+          `PluginLoader#loadPlugins: trying loading ${name}.${version}`
+        );
+        if (!version) return exports;
+
+        this.logger.debug(
+          `PluginLoader#loadPlugins: applying patch to ${name}@${version} using ${moduleName} module`
+        );
+
+        // Expecting a plugin from module;
+        try {
+          // tslint:disable-next-line:no-any
+          const plugin: Plugin<any> = require(moduleName).plugin;
+          this._plugins.push(plugin);
+          // Enable each supported plugin.
+          return plugin.enable(exports, this.tracer);
+        } catch (e) {
+          this.logger.error(
+            `PluginLoader#loadPlugins: could not load plugin ${moduleName} of module ${name}. Error: ${e.message}`
+          );
+          return exports;
+        }
+      });
+      this._hookState = HookState.ENABLED;
+    } else if (this._hookState === HookState.DISABLED) {
+      throw new Error(
+        'PluginLoader#loadPlugins: Currently cannot re-enable plugin loader.'
+      );
+    } else {
+      throw new Error(
+        'PluginLoader#loadPlugins: Plugin loader already enabled.'
+      );
+    }
+    return this;
+  }
+
+  /** Unloads plugins. */
+  unloadPlugins(): PluginLoader {
+    if (this._hookState === HookState.ENABLED) {
+      for (const plugin of this._plugins) {
+        plugin.disable();
+      }
+      this._plugins = [];
+      this._hookState = HookState.DISABLED;
+    }
+    return this;
+  }
+
+  /**
+   * Adds a search path for plugin modules. Intended for testing purposes only.
+   * @param searchPath The path to add.
+   */
+  static set searchPathForTest(searchPath: string) {
```

Comment:
Should this be an external utility instead of having it directly on the class if it's meant just for tests?

Reply:
AFAIK this can't be part of external utility, It's scoped to the specific file where it was called.

See https://nodejs.org/api/modules.html#modules_the_module_object: 
Since `require()` returns the `module.exports`, and the module is typically only available within a specific module's code, it must be explicitly exported in order to be used.

One possible solution, I can add utility in same module (which won't be directly on the class).

Reply:
Sorry if I was unclear, moving it out of the class is what I meant. It doesn't matter if it's in the same file or in a different file, it's just so that test functions are not exposed in the API.

Reply:
sure, done.